### PR TITLE
gcc: fix iostream init for Clang on Darwin

### PR DESCRIPTION
**Pull request by ds-saschasc**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This is a fix for gcc. Otherwise this will lead to an immediate segfault when including <iostream> in when using Clang in combination with libstdc++ 13.1. I've added a guard that the fix needs to be revisited when GCC 13.2 is released.

More details can be found here:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110432#c7

Here the fix which will be hopefully be included in GCC 13.2:
https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=fe2651affa8c15624188bfd062fb894648743431

There is no need to recreate the bundles. Therefore I suppose that the version number does not need to be increased.

### CODE REVIEW

File: Formula/gcc.rb
Comment (carlocab) [inline Formula/gcc.rb:244]: Let's not do this in `post_install`, please. Particularly since we are calling `odie`, this should happen as early in the install process as possible.

File: Formula/gcc.rb
Comment (carlocab) [inline Formula/gcc.rb:242]: We shouldn't hardcode `/usr/local` here, as that isn't the only prefix that we support.

**Review by carlocab — CHANGES_REQUESTED** [review]

**Comment by carlocab** [comment]
Thanks for the PR. I have a few comments, but we'll need approval from @fxcoudert before merging anything.

File: Formula/gcc.rb
Comment (saschasc) [reply Formula/gcc.rb:244]: Probably this could also be done differently without `odie`. I just wanted to assure that this fix gets removed for the GCC 13.2 release, since this should be fixed then.

Not sure why `post_install` is not good since this will have no side effect for GCC itself.

File: Formula/gcc.rb
Comment (saschasc) [reply Formula/gcc.rb:242]: `#{HOMEBREW_PREFIX}` would be the correct replacement for it?

File: Formula/gcc.rb
Comment (saschasc) [reply Formula/gcc.rb:244]: Would the bottles needed to be recreated if we put it in `install`? This is definitely not needed for the fix.

Comment (saschasc) [reply Formula/gcc.rb:244]: @carlocab To avoid the `odie` I could also change the if condition to `if version >= "13.1.0" && version < "13.2.0"

```
if version >= "13.1.0" && version < "13.2.0
  inreplace "/usr/local/opt/gcc@13/include/c++/13/iostream",
    /^(#if !__has_attribute\(__init_priority__\))$/, '\1 || defined __APPLE__'
end
```

Would this be better then? I thought to add the `odie` that this part gets detected more easily and then can be removed. The `odie` should  never happen to a normal user. Only (probably) when a maintainer updates to the GCC >= 13.2.

**Comment by fxcoudert** [comment]
The fix is fine, but we should:
- apply it at build-time (not postinstall)
- using the upstream patch

**Comment by m-a-v** [comment]
I had communicated with the developers of the corresponding branch which is referenced in this recipe and both the maintainer of this branch and a developer of libstdc++ found the approach ok. I don't know if it is worth to solve this "cleanly" for GCC 13.1. It will not make any difference.

https://github.com/iains/gcc-13-branch/issues/6#issuecomment-1615792059

@fxcoudert Would you or someone else take over?

**Comment by fxcoudert** [comment]
From the linked issue:

> As a matter of general policy I am trying to minimise the difference between these branches and upstream

So is Homebrew. We prefer to ship the patch itself. Could you amend your PR to do that?

**Comment by saschasc** [comment]
@fxcoudert Not sure if I will find enough time for it.

What would you expect. A separate patch which will be applied after:

https://github.com/Homebrew/homebrew-core/blob/ce44ea3d48ae3fead190de315558e271ee6b83e8/Formula/gcc.rb#L15

If not please communicate exactly how to realize it and if the `odie` would still be OK to review/remove the patch for GCC >= 13.2. I had seen that the corresponding place is even now slightly different from the version I have in Linux. There is a missing #elif statement in iostream.h which is irrelevant for macOS.

**Comment by fxcoudert** [comment]
The upstream commit might apply directly: https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff_plain;h=fe2651affa8c15624188bfd062fb894648743431;hp=d083c8c808cae2c7acb46117374bb047b4c3afce

If not, a patch to formula-patches, ping me on the PR.

`odie` is not needed: once the patch is not needed any more, applying it will fail if it is not removed.

**Comment by saschasc** [comment]
> The upstream commit might apply directly: https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff_plain;h=fe2651affa8c15624188bfd062fb894648743431;hp=d083c8c808cae2c7acb46117374bb047b4c3afce

@fxcoudert : So, I would copy the patch content to a file and then apply it after the line 16 in a separate patch command if possible? Correct?
https://github.com/Homebrew/homebrew-core/blob/ce44ea3d48ae3fead190de315558e271ee6b83e8/Formula/gcc.rb#L16

As far as I've seen it would be patcheable, but I will have to review the diffs.

I just want get exact instructions not to do it twice. Thanks.

**Comment by carlocab** [comment]
Something like this may work:
```patch
diff --git a/Formula/gcc.rb b/Formula/gcc.rb
index 87840c02dcb..04ccc7cc9e4 100644
--- a/Formula/gcc.rb
+++ b/Formula/gcc.rb
@@ -15,6 +15,12 @@ class Gcc < Formula
       url "https://raw.githubusercontent.com/Homebrew/formula-patches/5c206c47/gcc/gcc-13.1.0.diff"
       sha256 "cb4e8a89387f748a744da0273025d0dc2e3c76780cc390b18ada704676afea11"
     end
+
+    # Brief comment here explaining what this patch is for and when it can likely be removed.
+    patch do
+      url "https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff_plain;h=fe2651affa8c15624188bfd062fb894648743431;hp=d083c8c808cae2c7acb46117374bb047b4c3afce"
+      sha256 "353ccd8e2c53277f208ad5015e64d4031a6c08eefe5ae0ddbbad22a52a64e100"
+    end
   end
 
   livecheck do
```

**Comment by saschasc** [comment]
> If not, a patch to formula-patches, ping me on the PR.

 @fxcoudert It seems that the patch is not applicable without modifications. How to proceed?
 
```
patching file 'libstdc++-v3/acinclude.m4'
No such line 5679 in input file, ignoring
patching file 'libstdc++-v3/config.h.in'
patching file 'libstdc++-v3/configure'
patching file 'libstdc++-v3/configure.ac'
patching file 'libstdc++-v3/include/std/iostream'
patching file 'libstdc++-v3/src/c++98/ios_base_init.h'
```

**Comment by saschasc** [comment]
> If not, a patch to formula-patches, ping me on the PR.

```
patching file 'libstdc++-v3/acinclude.m4'
No such line 5679 in input file, ignoring
patching file 'libstdc++-v3/config.h.in'
patching file 'libstdc++-v3/configure'
patching file 'libstdc++-v3/configure.ac'
patching file 'libstdc++-v3/include/std/iostream'
patching file 'libstdc++-v3/src/c++98/ios_base_init.h'
```

The patch is not applicable from the existing diff file. This [diff](https://gcc.gnu.org/git/?p=gcc.git;a=blobdiff;f=libstdc%2B%2B-v3/acinclude.m4;h=823832f97d4f47d1161fe5e4d0dcd5c491305e8b;hp=277ae10e031a27c16354e3902a3bd7ea19780f53;hb=fe2651affa8c15624188bfd062fb894648743431;hpb=d083c8c808cae2c7acb46117374bb047b4c3afce) in the acinclude.m4 is ignored.

@fxcoudert How to proceed?

**Comment by saschasc** [comment]
Since I've not received any further feedback I will not invest more time into it. The fix will implicitly be included in the 13.2 release of gcc.

**Pull request closed without merging**
